File: src/main/java/ru/mrgrd56/api/controllers/UtilController.java
package ru.mrgrd56.api.controllers;

import org.apache.commons.lang3.ObjectUtils;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Objects;

@RestController
@RequestMapping("util")
public class UtilController {
    @RequestMapping(value = "delay", method = {RequestMethod.GET, RequestMethod.POST})
    public String delay(@RequestParam long ms) throws InterruptedException {
        Thread.sleep(ms);
        return Objects.toString(Math.random());
    }

    @RequestMapping(value = "http-response", method = {RequestMethod.GET, RequestMethod.POST})
    public ResponseEntity<?> getHttpResponse(
            @RequestParam(required = false, defaultValue = "200") Integer status,
            @RequestParam(required = false) List<String> header,
            @RequestParam(required = false) Object body) {
        HttpStatus responseStatus = status == null ? HttpStatus.OK : HttpStatus.valueOf(status);
        HttpHeaders responseHeaders = new HttpHeaders();

        if (header != null) {
            for (var rawHeader : header) {
                String[] parts = rawHeader.split(":\\n?", 2);
                if (parts.length != 2) {
                    return ResponseEntity.badRequest().body("Invalid header '" + rawHeader + "'");
                }

                responseHeaders.add(ObjectUtils.requireNonEmpty(parts[0]), parts[1]);
            }
        }

        return new ResponseEntity<>(body, responseHeaders, responseStatus);
    }
}

File: src/main/kotlin/ru/mrgrd56/api/controllers/UtilController.kt
package ru.mrgrd56.api.controllers

import freemarker.template.Configuration
import org.apache.commons.lang3.ObjectUtils
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import ru.mrgrd56.api.requestlogging.model.RequestLogDto
import ru.mrgrd56.api.requestlogging.services.RequestLoggingService
import ru.mrgrd56.api.utils.fill
import ru.mrgrd56.api.utils.logger
import java.util.*
import javax.servlet.http.HttpServletRequest

@RestController
@RequestMapping("util")
class UtilController(
    private val requestLoggingService: RequestLoggingService,
    private val freemarker: Configuration,
) {
    private val log = logger(this::class.java)

    @RequestMapping("delay", method = [RequestMethod.GET, RequestMethod.POST])
    @Throws(InterruptedException::class)
    fun delay(@RequestParam ms: Long): String {
        Thread.sleep(ms)
        return Objects.toString(Math.random())
    }

    @RequestMapping("http-response", method = [RequestMethod.GET, RequestMethod.POST])
    fun getHttpResponse(
        request: HttpServletRequest,
        @RequestParam(required = false, defaultValue = "200") status: Int?,
        @RequestParam(required = false) header: List<String>?,
        @RequestParam(required = false) body: Any,
        @RequestParam(required = false) log: String?,
        @RequestParam(name = "log.keepPersonal", defaultValue = "true") keepPersonal: Boolean,
        @RequestParam(name = "log.keepCredentials", defaultValue = "false") keepCredentials: Boolean,
        @RequestParam(name = "log.hideHeaders", defaultValue = "") hideHeaders: Set<String>
    ): ResponseEntity<*> {
        this.log.info("getHttpResponse: status={}, headers={}, body={}", status, header, body)

        if (!log.isNullOrBlank()) {
            try {
                requestLoggingService.logRequest(log, request, keepPersonal, keepCredentials, hideHeaders)
            } catch (e: Exception) {
                this.log.warn("getHttpResponse: Unable to log a request: {}", e.message, e)
            }
        }

        val responseStatus = status?.let { HttpStatus.valueOf(it) } ?: HttpStatus.OK
        val responseHeaders = HttpHeaders()

        header?.let {
            for (rawHeader in header) {
                val parts = rawHeader.split(":\\n?".toRegex(), limit = 2)
                if (parts.size != 2) {
                    return ResponseEntity.badRequest().body("Invalid header '$rawHeader'")
                }
                responseHeaders.add(ObjectUtils.requireNonEmpty(parts[0]), parts[1])
            }
        }

        return ResponseEntity(body, responseHeaders, responseStatus)
    }

    @RequestMapping("log-request/{loggerId}")
    fun logRequest(request: HttpServletRequest,
                   @PathVariable loggerId: String,
                   @RequestParam(defaultValue = "true") keepPersonal: Boolean,
                   @RequestParam(defaultValue = "false") keepCredentials: Boolean,
                   @RequestParam(defaultValue = "") hideHeaders: Set<String>) {
        requestLoggingService.logRequest(loggerId, request, keepPersonal, keepCredentials, hideHeaders)
    }

    @GetMapping("log-request/{loggerId}/logs.json")
    fun getLoggedRequests(@PathVariable loggerId: String): List<RequestLogDto> {
        return requestLoggingService.getLoggedRequests(loggerId)
    }

    @GetMapping("log-request/{loggerId}/logs",
        produces = [MediaType.TEXT_HTML_VALUE]
    )
    fun viewLoggedRequests(@PathVariable loggerId: String): String {
        val template = freemarker.getTemplate("request-logs.html.ftl")

        return template.fill(mapOf(
            "requests" to requestLoggingService.getLoggedRequests(loggerId)
        ))
    }

    @RequestMapping("log-request/{loggerId}/clear",
        method = [RequestMethod.GET, RequestMethod.POST, RequestMethod.PUT, RequestMethod.DELETE]
    )
    fun clearLoggerRequests(@PathVariable loggerId: String) {
        requestLoggingService.clearLoggedRequests(loggerId)
    }
}
